src/api/providers/cline.ts

```diff
@@ -1,143 +1,233 @@
-					throw new Error("Cline Mid-Stream Error: Stream terminated with error status but no error details provided")
-				}
+			if (!response.data.data.choices || response.data.data.choices.length === 0) {
+				throw new Error(`No choices returned from Cline API: ${JSON.stringify(response.data)}`)
 			}
 
-			const delta = choice?.delta
-			if (delta?.content) {
-				yield {
-					type: "text",
-					text: delta.content,
+			for (const choice of response.data.data.choices) {
+				if (choice.finish_reason === "error") {
+					const error = choice.error || { code: "Unknown", message: "No error details provided" }
+					console.error(`Cline API Error: ${error.code} - ${error.message}`)
+					throw new Error(`Cline API Error: ${error.code} - ${error.message}`)
 				}
-			}
-
-			// Reasoning tokens are returned separately from the content
-			if ("reasoning" in delta && delta.reasoning) {
-				yield {
-					type: "reasoning",
-					// @ts-ignore-next-line
-					reasoning: delta.reasoning,
+				if (choice.delta && choice.delta.content) {
+					yield {
+						type: "text",
+						text: choice.delta.content,
+					}
 				}
-			}
-
-			if (!didOutputUsage && chunk.usage) {
-				// @ts-ignore-next-line
-				let totalCost = (chunk.usage.cost || 0) + (chunk.usage.cost_details?.upstream_inference_cost || 0)
-				const modelId = this.getModel().id
-				const provider = modelId.split("/")[0]
-
-				// If provider is x-ai, set totalCost to 0 (we're doing a promo)
-				if (provider === "x-ai") {
-					totalCost = 0
+				if (choice.delta && choice.delta.reasoning) {
+					yield {
+						type: "reasoning",
+						reasoning: choice.delta.reasoning,
+					}
 				}
-
-				if (modelId.includes("gemini")) {
+				if (choice.message && choice.message.content) {
 					yield {
-						type: "usage",
-						cacheWriteTokens: 0,
-						cacheReadTokens: chunk.usage.prompt_tokens_details?.cached_tokens || 0,
-						inputTokens: (chunk.usage.prompt_tokens || 0) - (chunk.usage.prompt_tokens_details?.cached_tokens || 0),
-						outputTokens: chunk.usage.completion_tokens || 0,
-						// @ts-ignore-next-line
-						totalCost,
+						type: "text",
+						text: choice.message.content,
 					}
-				} else {
+				}
+				if (choice.usage) {
+					const totalCost = choice.usage.cost || 0
 					yield {
 						type: "usage",
 						cacheWriteTokens: 0,
-						cacheReadTokens: chunk.usage.prompt_tokens_details?.cached_tokens || 0,
-						inputTokens: chunk.usage.prompt_tokens || 0,
-						outputTokens: chunk.usage.completion_tokens || 0,
-						// @ts-ignore-next-line
+						cacheReadTokens: choice.usage.cached_tokens || 0,
+						inputTokens: choice.usage.prompt_tokens || 0,
+						outputTokens: choice.usage.completion_tokens || 0,
 						totalCost,
 					}
 				}
+			}
+
+			if (response.data.data.usage) {
 				didOutputUsage = true
+				yield {
+					type: "usage",
+					cacheWriteTokens: 0,
+					cacheReadTokens: response.data.data.usage.prompt_tokens_details.cached_tokens || 0,
```

Comment:
Is there a possibility that `prompt_tokens_details` is undefined, in which case we should use an optional `?` after? Otherwise we would get a type error trying to access `cached_tokens` on an undefined value.

Reply:
Good Call, Updating